src/modules/linux-micro/kmod/kmod.c

```diff
@@ -0,0 +1,475 @@
+        kmod_apply_value(kdata->kmod, modalias);
+    }
+}
+
+static bool
+kmod_uevent_handler(void *data, int fd, unsigned int cond)
+{
+    int len;
+    struct kmod_data *kdata = data;
+    char buffer[512];
+
+    len = recv(kdata->uevent.fd, buffer, sizeof(buffer), MSG_WAITALL);
+    if (len == -1) {
+        SOL_ERR("Could not read netlink socket. %s", sol_util_strerrora(errno));
+        return false;
+    } else {
+        uevent_read_msg(kdata, buffer, len);
+    }
+
+    return true;
+}
+
+static int
+kmod_uevent_start(struct kmod_data *kdata)
+{
+    struct sockaddr_nl nls;
+
+    memset(&nls, 0, sizeof(struct sockaddr_nl));
+    nls.nl_family = AF_NETLINK;
+    nls.nl_pid = getpid();
+    nls.nl_groups = -1;
+
+    kdata->uevent.fd = socket(PF_NETLINK, SOCK_DGRAM | SOCK_CLOEXEC,
+        NETLINK_KOBJECT_UEVENT);
+    if (kdata->uevent.fd == -1) {
+        SOL_ERR("Could not open uevent netlink socket.");
+        return -errno;
+    }
+
+    if (bind(kdata->uevent.fd, (void *)&nls, sizeof(struct sockaddr_nl))) {
+        SOL_ERR("Could not bind to uevent socket.");
+        close(kdata->uevent.fd);
+        kdata->uevent.fd = 0;
+        return -errno;
+    }
+
+    kdata->uevent.watch = sol_fd_add(kdata->uevent.fd, SOL_FD_FLAGS_IN,
+        kmod_uevent_handler, kdata);
+    return 0;
+}
+
+static void
+kmod_coldplug_find_devices(struct kmod_data *kdata, DIR *dir)
+{
+    struct dirent ent, *entbuf = NULL;
+    int r, dfd;
+
+    dfd = dirfd(dir);
+    if (dfd < 0) {
+        SOL_DBG("Could not obtain directory file descriptor");
+        return;
+    }
+
+    r = readdir_r(dir, &ent, &entbuf);
+    while (r == 0 && entbuf) {
+        struct stat st;
+
+        if (!strcmp(ent.d_name, ".") || !strcmp(ent.d_name, "..")) {
+            goto next;
+        }
+
+        if (fstatat(dfd, ent.d_name, &st, AT_SYMLINK_NOFOLLOW) == -1) {
+            SOL_ERR("Could not stat %s", ent.d_name);
+            goto next;
+        }
+
+        if (S_ISDIR(st.st_mode)) {
+            int nfd;
+            DIR *ndir;
+
+            nfd = openat(dfd, ent.d_name, O_RDONLY | O_CLOEXEC);
+            if (nfd < 0) {
+                SOL_ERR("Could not open sys directory %s - %s", ent.d_name,
+                    sol_util_strerrora(errno));
+                goto next;
+            }
+
+            ndir = fdopendir(nfd);
+            if (!ndir) {
+                SOL_ERR("Could not open sys directory %s", ent.d_name);
+                close(nfd);
+                goto next;
+            }
+
+            kmod_coldplug_find_devices(kdata, ndir);
+            closedir(ndir);
+        } else if (!strcmp(ent.d_name, "modalias")) {
+            int fd;
+            char buff[PATH_MAX];
+
+            fd = openat(dfd, ent.d_name, O_RDONLY | O_CLOEXEC);
+            if (fd < 0) {
+                SOL_ERR("Could not obtain uevent file descriptor");
+                goto next;
+            }
+
+            if (read(fd, buff, sizeof(buff)) > 0) {
```

Comment:
Why not use `sol_file_reader` here, that will handle `EINTR`, `EAGAIN` and whatnot?